### PR TITLE
KAFKA-7192: Wipe out if EOS is turned on and checkpoint file does not exist

### PR DESCRIPTION
**Pull request by guozhangwang**
1. As titled and as described in comments.
2. Modified unit test slightly to insert for new keys in committed data to expose this issue.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@mjsax @bbejeck @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java [hunk 1]
```diff
@@ -186,8 +187,24 @@ private void startRestoration(final Map<TopicPartition, StateRestorer> initializ
...
+                if (task.isEosEnabled()) {
+                    log.info("No checkpoint found for task {} state store {} changelog {} with EOS turned on. " +
+                            "Reinitializing the task.", task.id, restorer.storeName(), restorer.partition());
+                    task.reinitializeStateStoresForPartitions(Collections.singleton(restorer.partition()));
+                    restoreConsumer.seekToBeginning(Collections.singleton(restorer.partition()));
+
+                } else {
+                    log.info("Restoring task {}'s state store {} from beginning of the changelog {} ", task.id, restorer.storeName(), restorer.partition());
+
+                    restoreConsumer.seekToBeginning(Collections.singletonList(restorer.partition()));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java]: We `seekToBeginning` in both cases -- should we extract and share this?

Also, even for non-EOS, what would be the disadvantage in wiping out the store, too? We could still have some writes to the store that might not be cleaned up, and it seems we can get this cleanup for free?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java [hunk 2]
```diff
@@ -631,16 +628,29 @@ public void init(final ProcessorContext context) {
...
                                 sum = value;
                             } else {
                                 sum += value;
                             }
                             state.put(key, sum);
-                            context.forward(key, sum);
-                            return null;
+                            state.flush();
+                            //context.forward(key, sum);
+                            //return null;
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java]: nit: should be removed?

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java]: ack

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java]: We had this discussion before :) the concern is that wiping out the store is not completely free: 1) you call `store.close()` and then 2) delete the corresponding directory and then 3) recreate the directory with the state again. Among them 1) and 3) may take time. And the rationale is that if EOS is not turned on then not clean up the existing writes but just let them be overwritten from the changelog (note there may be duplicates, hence at-least-once) is fine.

**Review by guozhangwang — COMMENTED** [review]
Addressed comments.

**Comment by mjsax** [comment]
Seems like, we should cherry-pick to `2.0` and maybe also to `1.1`? (Including follow up PR #5430)

WDYT?

**Comment by guozhangwang** [comment]
Will try to do that now.

**Comment by guozhangwang** [comment]
Cherry-picked to 2.0

**Pull request merged**